### PR TITLE
fx quant: refactor observer insertion

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57519 fx quant: refactor observer insertion**

Summary:

tl;dr; rewrites the FX graph mode quantization observer insertion to be easier to understand and extend.
The key conceptual difference from before is:
* before: for each node, observers are always inserted to the output of the current node, even if they are needed for the next node. This is hard to reason about.
* after: for each node, observers are inserted to the inputs (if needed, as calculated by the dtype of the argument and dtype of current node) and to the output (if needed for the type of pattern and qconfig).  There is no knowledge of future nodes needed to insert observers for the current node.

This allows us to significantly simplify various things:
* all new observers needed for a node are inserted together.  This makes it easier to understand and debug things.  We add an invariant that node X will never change any observers inserted by any preceding or subsequent node, so to debug an issue the user can just understand what is happening for node X, without having to understand what happens before or after it.
* all the state tracking of activation_post_process_map and activation_post_process_indices are removed, instead observers are looked up by graph traversals
* since there is no longer a need for overlapping graph passes which mutate each other's interemediate state, it is easier to understand what the rules are for inserting observers, and to create new rules in the future.

Test Plan:

```
# all OSS tests pass
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28241864](https://our.internmc.facebook.com/intern/diff/D28241864)

### CODE REVIEW

File: torch/quantization/fx/quantization_patterns.py [hunk 1]
```diff
@@ -67,6 +67,27 @@ def __init__(self, quantizer: QuantizerCls, node: Node):
...
+        quantizer: QuantizerCls,
+    ) -> Optional[Union[torch.quantization.ObserverBase, torch.quantization.FakeQuantizeBase]]:
+        """
+        If the last node of the pattern is observed, return the observer
+        instance. Otherwise, return None.
+        """
+        for maybe_obs_node, _ in self.last_node.users.items():
+            if maybe_obs_node.op == 'call_module':
+                maybe_obs = quantizer.modules[maybe_obs_node.name]
+                if (
```
Comment (jerryzh168) [inline torch/quantization/fx/quantization_patterns.py:84]: nit: we have: https://github.com/pytorch/pytorch/blob/master/torch/quantization/quantize.py#L23

File: torch/quantization/fx/quantization_patterns.py [hunk 2]
```diff
@@ -288,10 +309,14 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
                 args = load_arg(quantized=False)(node.args)
                 kwargs = load_arg(quantized=False)(node.kwargs)
                 op_out = quantizer.quantized_graph.node_copy(node, load_arg(quantized=False))
-                cur_idx = quantizer.activation_post_process_indexes[node.name]
                 activation_post_process = \
-                    quantizer.modules[quantizer.activation_post_process_map[node.name][cur_idx]]
-                quantizer.activation_post_process_indexes[node.name] += 1
+                    self._maybe_get_last_node_only_observer(quantizer)
+                assert activation_post_process is not None
```
Comment (jerryzh168) [inline torch/quantization/fx/quantization_patterns.py:314]: should we make this assert in _maybe_get_last_node_only_observer?

Comment (vkuzo) [reply torch/quantization/fx/quantization_patterns.py:314]: hmm, so this is possible at the individual quantize_handler class level, but not possible for the base class (because not all quantize_handlers are expected to have output be observed).

Comment (jerryzh168) [reply torch/quantization/fx/quantization_patterns.py:314]: Oh OK

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28241864).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28241864).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28241864).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28241864).

File: torch/quantization/fx/quantize.py [hunk 3]
```diff
@@ -95,10 +95,13 @@
 MatchResult = Tuple[Node, List[Node], Optional[Pattern], QuantizeHandler,
                     QConfigAny]
 
+DType = Any  # torch.float, torch.quint8, etc
```
Comment (jerryzh168) [inline torch/quantization/fx/quantize.py:98]: nit: does torch.dtype work?

File: torch/quantization/fx/quantize.py [hunk 4]
```diff
@@ -406,6 +415,606 @@ def load_arg(a):
...
+            observer = qconfig.activation()
+            new_obs = insert_observer_rewrite(
+                prev_node, observer, model, modules, graph)
+            # set the type, so the next node can read it
+            node_name_to_target_dtype[new_obs.name] = node_dtype
+            return new_obs
+
+    return None
+
+def adjust_observers_for_cat_rewrite(
```
Comment (vkuzo) [inline torch/quantization/fx/quantize.py:772]: note: this does not work well yet, need to update

File: torch/quantization/fx/quantize.py (hunk 3 above)
Comment (vkuzo) [reply torch/quantization/fx/quantize.py:98]: thanks, will try this

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28241864).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28241864).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28241864).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28241864).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28241864).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28241864).

File: torch/quantization/fx/quantize.py [hunk 5]
```diff
@@ -407,6 +414,687 @@ def load_arg(a):
...
+            node.name, (None, None, None, None, None))
+        node_name_to_target_dtype[node.name] = get_target_dtype_for_node_rewrite(
+            node, qconfig, inputs_seen_counter, outputs_seen_counter,
+            input_quantized_idxs, output_quantized_idxs, qhandler)
+
+    # Second, for nodes with known input dtypes, propagate them throughout the
+    # graph. For example, if there is a call such as
+    #   x1 = x0.masked_fill(mask, 1)
+    # we propagate the type of mask to be torch.bool
+    propagate_dtypes_for_known_nodes_rewrite(
```
Comment (jerryzh168) [inline torch/quantization/fx/quantize.py:940]: should we incorporate all_node_args_have_no_tensors here as well? basically we are assuming all nodes are Tensors here, we can use None as a dtype for non-Tensors?

File: torch/quantization/fx/quantize.py [hunk 6]
```diff
@@ -407,6 +414,687 @@ def load_arg(a):
...
+            # if a graph input is in int8, we assume the observation happens
+            #   outside of the graph, and no additional observation is needed
+            pass
+
+        elif new_node.op in ('call_module', 'call_method', 'call_function', 'output'):
+            modules = dict(model.named_modules(remove_duplicate=False))
+            args_have_no_tensors = \
+                all_node_args_have_no_tensors(
+                    new_node, modules, cache_for_no_tensor_check)
+            is_getitem = new_node.op == 'call_function' and \
```
Comment (jerryzh168) [inline torch/quantization/fx/quantize.py:976]: this would be equivalent to removing operator.getitem in the quantization_patterns.py? does the test_general_shape_ops work?

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28241864).

File: torch/quantization/fx/quantize.py (hunk 6 above)
Comment (vkuzo) [reply torch/quantization/fx/quantize.py:985]: the `test_general_shape_ops` test passes.  Removing operator.getitem from patterns sounds good to me, maybe that can be done in a future PR, just to keep this PR as a refactor with minimal logic changes?

File: torch/quantization/fx/quantize.py [hunk 7]
```diff
@@ -407,6 +414,696 @@ def load_arg(a):
...
+    first_arg_arg = first_arg[0]
+
+    # if we have a graph such as
+    #   observed_node -> non_observed_node -> cat
+    # we need to navigate up to the first observer
+    iteration_guard = 0
+    while not is_activation_post_process_node(first_arg_arg, modules):
+        first_arg_arg = first_arg_arg.args[0]
+        iteration_guard += 1
+        if iteration_guard > 10:
```
Comment (supriyar) [inline torch/quantization/fx/quantize.py:852]: What is the reason to stop at 10 here?

Comment (supriyar) [reply torch/quantization/fx/quantize.py:852]: Will this fail in case we have > 10 non_observed nodes between the first observer and current op? Wondering if that is a valid case we expect to see in models.

**Comment by supriyar** [comment]
Looks great! Do we plan to have a test case that checks that we insert observers around the use (not the source)
Example quantized graph of this case

# current
                  int8_op
                    /
x0 -> quant -> dequant -> fp32_op

# expected
        
x0 -> quant -> int8_op
 \
 fp32_op

File: torch/quantization/fx/quantize.py (hunk 7 above)
Comment (vkuzo) [reply torch/quantization/fx/quantize.py:852]: sure, we can increase the limit to 10e5

**Comment by vkuzo** [comment]
> Looks great! Do we plan to have a test case that checks that we insert observers around the use (not the source)
> Example quantized graph of this case
> 
> current
> 
> ```
>            int8_op
>              /
> x0 -> quant -> dequant -> fp32_op
> ```
> 
> expected
> 
> ```
> x0 -> quant -> int8_op
> \
> fp32_op
> ```

sure.  This logic to do this had to be added in order for existing tests to pass, but we can also add a new test to test explicitly for this.

File: torch/quantization/fx/quantize.py [hunk 8]
```diff
@@ -407,6 +414,696 @@ def load_arg(a):
...
+                        # this modifies new_node inplace
+                        maybe_insert_input_observers_for_node_rewrite(
+                            new_node, prev_node_qconfig, load_arg, model,
+                            modules, observed_graph,
+                            node_name_to_target_dtype,
+                            qhandler, prepare_custom_config_dict)
+            else:
+                # if we skipped observation and we have a copy node,
+                # assign the dtype to the dtype of the previous node
+                # new_node.dtype = new_node.args[0].dtype
```
Comment (jerryzh168) [inline torch/quantization/fx/quantize.py:1048]: is this comment up to date? not sure what this means. is this addressed before alraedy?

Comment (vkuzo) [reply torch/quantization/fx/quantize.py:1048]: sorry, this is stale, we can remove the comment and move the code to mark 
`getitem` outputs as `torch.float` to `get_target_dtype_for_node_rewrite`

File: torch/quantization/fx/quantize.py [hunk 9]
```diff
@@ -407,6 +414,696 @@ def load_arg(a):
...
+                        (qhandler is not None and (
+                            isinstance(qhandler, CopyNodeQuantizeHandler)
+                        ))
+                    if is_last_node_of_pattern and (not is_like_copy_node):
+                        # this returns the new observer node if it was needed
+                        maybe_output_obs_node = maybe_insert_output_observer_for_node_rewrite(
+                            new_node, model, modules, observed_graph, matches,
+                            node_name_to_target_dtype, pattern, qhandler)
+                        if maybe_output_obs_node is not None:
+                            env[new_node.name] = maybe_output_obs_node
```
Comment (jerryzh168) [inline torch/quantization/fx/quantize.py:1013]: looks like we updated the value for new_node.name here, is supriya's comment address in the PR? https://github.com/pytorch/pytorch/pull/57519#issuecomment-840094762

File: torch/quantization/fx/quantize.py [hunk 10]
```diff
@@ -407,6 +414,696 @@ def load_arg(a):
...
+    inputs_seen_counter: int,
+    outputs_seen_counter: int,
+    input_quantized_idxs: List[int],
+    output_quantized_idxs: List[int],
+    qhandler: Optional[QuantizeHandler],
+    modules: Dict[str, torch.nn.Module],
+    cache_for_no_tensor_check: Dict[Node, bool],
+) -> Optional[torch.dtype]:
+    """
+    Returns the expected dtype of the input and output of this node after
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:479]: +1 for this, might be better to mention this in name or comment, since the weight is handled in a different place

File: torch/quantization/fx/quantize.py [hunk 11]
```diff
@@ -407,6 +414,696 @@ def load_arg(a):
...
+            new_inner_arg = maybe_insert_input_observer_for_arg_or_kwarg_rewrite(
+                node, inner_arg, qconfig, load_arg, model, modules,
+                graph, node_name_to_target_dtype,
+                qhandler, prepare_custom_config_dict)
+            new_arg_to_return.append(new_inner_arg)
+        return new_arg_to_return
+
+    if not isinstance(arg, Node):
+        return arg
+    assert isinstance(arg, Node)
```
Comment (jerryzh168) [inline torch/quantization/fx/quantize.py:559]: nit: is this a bit redundant?

File: torch/quantization/fx/quantize.py (hunk 9 above)
Comment (vkuzo) [reply torch/quantization/fx/quantize.py:1013]: yes, this should work without changes.  The key distinction is this only applies to node output, and a node can only have a single output dtype of the first argument after convert.  

for example,

```
x0_fp32 -> obs0 -> int8_op
\
fp32_op
```

in the graph above, x0_fp32's target dtype is fp32, so no observer will be inserted to its output.  `obs0` will be inserted when the `int8_op` node is processed.  This is a key difference of the proposal in this PR (each node inserts input and output observers it needs) versus the previous behavior (observers are always inserted to the output of current node, even if they are only needed for a future node)

File: torch/quantization/fx/quantize.py (hunk 11 above)
Comment (vkuzo) [reply torch/quantization/fx/quantize.py:559]: yeah, this is just for making mypy happy.  I'll double check if I can remove it.

File: torch/quantization/fx/quantize.py (hunk 9 above)
Comment (vkuzo) [reply torch/quantization/fx/quantize.py:1013]: or, to explain it in a different way:
* a node can only have one output type after convert
* the "multiple observers of a single node" use case comes from nodes of different target dtypes having a shared node as an input
* the observers for inputs of a node are not handled in this code path (and they are inserted without modifying `env`)

File: torch/quantization/fx/quantize.py [hunk 12]
```diff
@@ -407,6 +414,696 @@ def load_arg(a):
...
+        if existing_obs_node is None:
+            new_obs_node = insert_observer_rewrite(
+                arg_this_env, new_obs_mod, model, modules, graph)
+            # set the type, so the next node can read it
+            node_name_to_target_dtype[new_obs_node.name] = node_dtype
+            # override this arg to be the observed arg
+            new_arg = new_obs_node
+        else:
+            new_arg = existing_obs_node
+
```
Comment (jerryzh168) [inline torch/quantization/fx/quantize.py:657]: if it does not need observer, looks like we would return `arg`, which is from the old graph?

Comment (vkuzo) [reply torch/quantization/fx/quantize.py:657]: correct, this part of the code mutates the nodes of an existing graph (without creating a new graph).

if the RFC goes forward, a potential future improvement would be to stop using the `env` pattern in inserting observers for outputs as well, and mutate the graph directly.  Using env works ok for inserting observers to outputs (because we are iterating through the graph in order), but it does not work cleanly for inserting observers to node inputs (since this is out of order).

Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:657]: oh I see, yeah having env and mutating the graph sounds a bit confusing

**Review by jerryzh168 — APPROVED** [review]
Looks good, thanks for the refactor! I think the logic is definitely cleaner than before.

One concern I have is about using both mutation and `env` in the code, I feel it would be even clearer if we only use `env` and remove the mutation. For that to happen we need to make env to be Dict[Tuple[str, torch.dtype], Node] to record different dtypes for the same node.

Another comment I have is that we may need to consider is what to do in this case:
qconfig:
```python
op1 (quint8) -> op2 (fp16)
   \ ----- op3 (quint8)
```
Option 1:
```python

                                         /--> op3(quint8) --> quantize_per_tensor
qdq--- op1 (quint8) -quantize_per_tensor(quint8) -> dequant
                                                        \----> to(torch.float16) -> dequant -> op2 (fp16) -> to(torch.float16)
```
that is we insert a quantize_per_tensor for output of op1 and use it as the input for both op2 and op3, or

Option 2:
```python

                                /--dequant --> op3(quint8) --> quantize_per_tensor --
-qdq-- op1 (quint8) -quantize_per_tensor(quint8)  
                   \--------> to(torch.float16) -> dequant -> op2 (fp16) -> to(torch.float16)
```
that is op2 is getting the input from original floating point output of op1 directly, instead of the quantized one.

I know current behavior is the first option, but wondering if we'll have a case when the second one makes more sense.

**Comment by vkuzo** [comment]
> Another comment I have is that we may need to consider is what to do in this case:

oh, sorry, the ASCII graphs you specified did not render properly for some reason, it's hard to see what connects to what.  Maybe we can create a gist with them?  I also noticed they render incorrectly in my previous PRs, maybe github UI rendering has a bug.

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28241864).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28241864).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28241864).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28241864).

**Comment by vkuzo** [comment]
this is closed and merged
